core/tx_pool.go

```diff
@@ -1590,11 +1592,49 @@ func (pool *TxPool) queueTxEvent(tx *types.Transaction) {
 
 // SendTxToSharingClients sends the tx into the pool sharing tx ch and
 // if its full logs it
-func (pool *TxPool) SendTxToSharingClients(tx *types.Transaction) {
-	select {
-	case pool.poolSharingTxCh <- tx:
-	default:
-		pool.logger.Warn("pool sharing tx ch is full")
+func (pool *TxPool) SendTxToSharingClients(tx *types.Transaction) error {
+	pool.sharingClientMu.Lock()
+	defer pool.sharingClientMu.Unlock()
+
+	// If there are no tx pool sharing clients just submit to the local pool
+	if len(pool.config.SharingClientsEndpoints) == 0 || !pool.config.SyncTxWithReturn {
+		err := pool.AddLocal(tx)
+		if err != nil {
+			return err
+		}
+		select {
+		case pool.poolSharingTxCh <- tx:
+		default:
+			pool.logger.Warn("pool sharing tx ch is full")
+		}
+		return err
+	} else {
+		// send to the first client, and then submit to the rest
+		client := pool.poolSharingClients[0]
+		ctx, cancel := context.WithTimeout(context.Background(), txSharingPoolTimeout)
+		defer cancel()
+		err := client.SendTransactionToPoolSharingClient(ctx, tx)
+		if err != nil {
+			pool.logger.WithField("err", err).Error("Error sending transaction to pool sharing client")
+		}
+
+		if len(pool.poolSharingClients) > 1 {
+			// send to all pool sharing clients
+			for _, client := range pool.poolSharingClients[1:] {
+				if client != nil {
+					go func(*quaiclient.Client, *types.Transaction) {
+						ctx, cancel := context.WithTimeout(context.Background(), txSharingPoolTimeout)
+						defer cancel()
+						sendErr := client.SendTransactionToPoolSharingClient(ctx, tx)
```

Comment:
Previously this was done in a goroutine, I am not arguing one way or another but are we sure this is the better approach? Why did we use goroutines initially?

Reply:
go routine is so that we dont block, and we have a deadline with timeout, so we wont have leaks, but we wont block anything, its becomes a problem when we have many nodes to service, we cant do it in sequence